src/ivectorbin/ivector-plda-scoring-dense.cc

```diff
@@ -0,0 +1,223 @@
+      &transformed_ivector);
+    ivectors_out->Row(i).CopyFromVec(transformed_ivector);
+  }
+}
+
+// Transform the i-vectors using the recording-dependent PCA matrix.
+void ApplyPca(const Matrix<BaseFloat> &ivectors_in,
+  const Matrix<BaseFloat> &pca_mat, Matrix<BaseFloat> *ivectors_out) {
+  int32 transform_cols = pca_mat.NumCols(),
+        transform_rows = pca_mat.NumRows(),
+        feat_dim = ivectors_in.NumCols();
+  ivectors_out->Resize(ivectors_in.NumRows(), transform_rows);
+  KALDI_ASSERT(transform_cols == feat_dim);
+  ivectors_out->AddMatMat(1.0, ivectors_in, kNoTrans,
+    pca_mat, kTrans, 0.0);
+}
+
+} // namespace kaldi
+
+int main(int argc, char *argv[]) {
+  using namespace kaldi;
+  typedef kaldi::int32 int32;
+  try {
+    const char *usage =
+      "Perform PLDA scoring for speaker diarization.  The input reco2utt\n"
+      "should be of the form <recording-id> <seg1> <seg2> ... <segN> and\n"
+      "there should be one iVector for each segment.  PLDA scoring is\n"
+      "performed between all pairs of iVectors in a recording and outputs\n"
+      "an archive of score matrices, one for each recording-id.  The rows\n"
+      "and columns of the the matrix correspond the sorted order of the\n"
+      "segments.\n"
+      "Usage: ivector-plda-scoring-dense [options] <plda> <reco2utt>"
+      " <ivectors-rspecifier> <scores-wspecifier>\n"
+      "e.g.: \n"
+      "  ivector-plda-scoring-dense plda reco2utt scp:ivectors.scp"
+      " ark:scores.ark ark,t:ivectors.1.ark\n";
+
+    ParseOptions po(usage);
+    BaseFloat target_energy = 0.5;
+    PldaConfig plda_config;
+    plda_config.Register(&po);
+
+    po.Register("target-energy", &target_energy,
+      "Reduce dimensionality of i-vectors using a recording-dependent"
+      " PCA such that this fraction of the total energy remains.");
+    KALDI_ASSERT(target_energy <= 1.0);
+
+    po.Read(argc, argv);
+
+    if (po.NumArgs() != 4) {
+      po.PrintUsage();
+      exit(1);
+    }
+
+    std::string plda_rxfilename = po.GetArg(1),
+      reco2utt_rspecifier = po.GetArg(2),
+      ivector_rspecifier = po.GetArg(3),
+      scores_wspecifier = po.GetArg(4);
+
+    Plda plda;
+    ReadKaldiObject(plda_rxfilename, &plda);
+
+    SequentialTokenVectorReader reco2utt_reader(reco2utt_rspecifier);
+    RandomAccessBaseFloatVectorReader ivector_reader(ivector_rspecifier);
+    BaseFloatMatrixWriter scores_writer(scores_wspecifier);
+    int32 num_spk_err = 0,
+          num_spk_done = 0;
+    for (; !reco2utt_reader.Done(); reco2utt_reader.Next()) {
+      Plda this_plda(plda);
+      std::string reco = reco2utt_reader.Key();
+
+      // The uttlist is sorted here and in binaries that use the scores
+      // this outputs.  This is to ensure that the segment corresponding
+      // to the same rows and columns (of the score matrix) across binaries.
+      std::vector<std::string> uttlist = reco2utt_reader.Value();
+      std::sort(uttlist.begin(), uttlist.end());
```

Comment:
Do you really need to sort here? It might be better for diagnostics if you just use the same order of utterances as in reco2utt. Since reco2utt is being copied across directories, the same order will be ensured in different programs.